src/roll.c

```diff
@@ -0,0 +1,143 @@
+
+  nx = length(x);
+  i = 0;                                                                  // check that every column is double/integer type
+  while (i < nx && (isReal(VECTOR_ELT(x, i)) || isInteger(VECTOR_ELT(x, i)))) i++;
+  if (i != nx)
+    error("x must be list, data.frame or data.table of numeric types");
+
+  nk = length(k);
+  if (nk == 0)                                                            // check that window is not empty
+    error("n must be non 0 length");
+
+  if (!isLogical(adaptive) || length(adaptive) != 1 || LOGICAL(adaptive)[0] == NA_LOGICAL)
+    error("adaptive must be logical TRUE or FALSE");
+
+  if (!LOGICAL(adaptive)[0]) {                                            // validating n input for adaptive=FALSE
+    if (isNewList(k))
+      error("n must be integer, list is accepted for adaptive TRUE");
+    //TODO move R's n=as.integer(n) to C
+    if (!isInteger(k))                                                    // check that k is integer vector
+      error("n must be integer");
+    i = 0;                                                                // check that all window values non-negative
+    while (i < nk && INTEGER(k)[i] >= 0) i++;
+    if (i != nk)
+      error("n must be non-negative integer values (>= 0)");
+  } else if (LOGICAL(adaptive)[0]) {                                      // validating n input for adaptive=TRUE
+    if (isVectorAtomic(k)) {                                              // if not-list then wrap into list
+      kl = allocVector(VECSXP, 1);
+      SET_VECTOR_ELT(kl, 0, k);
+      nk = 1;
+    } else {
+      kl = k;
+    }
+    //TODO move R's n=as.integer(n) to C
+    i = 0;                                                                // check that every column is integer type
+    while (i < nk && isInteger(VECTOR_ELT(kl, i))) i++;
+    if (i != nk)
+      error("n must be list of integer vectors when adaptive TRUE");
+  }
+  
+  if (length(fill) != 1)
+    error("fill must be a vector of length 1");
+
+  if (!strcmp(CHAR(STRING_ELT(align, 0)), "right"))
+    salign = RIGHT;
+  else if (!strcmp(CHAR(STRING_ELT(align, 0)), "center"))
+    salign = CENTER;
+  else if (!strcmp(CHAR(STRING_ELT(align, 0)), "left"))
+    salign = LEFT;
+  else
+    error("Internal error: invalid align argument in rolling function, should have been caught before. please report to data.table issue tracker.");
+
+  if (!isLogical(narm) || length(narm)!=1 || LOGICAL(narm)[0]==NA_LOGICAL)
+    error("na.rm must be TRUE or FALSE");
+  
+  ans = PROTECT(allocVector(VECSXP, nk * nx)); protecti++;                // allocate list to keep results
+  
+  thisfill = PROTECT(coerceVector(fill, REALSXP)); protecti++;
+  
+  if (!LOGICAL(adaptive)[0]) {                                          // adaptive=FALSE
+    if (salign == RIGHT) {                                              // align right scenario
+      for (i=0; i<nx; i++) {                                            // loop over columns
```

Comment:
We could merge two for loops into one, so we can parallelize on columns and on window sizes, not just on one of them.

Reply:
Good thought. But this leaves too much up to OpenMP, I've found.  My current thought.  To date, I prefer parcelling up the work myself and having more loops, in fact.  In the 2-loop version above, we can look at the j=0..nk loop and reason with that. We know one thread will do that exclusively.  Automatic variables inside that loop definition will be private to that thread, with no sharing.   You can't do that sort of reasoning when there's just one big fat loop handed over to OpenMP.  There's no way to stop the programmer doing that, but it isn't the best way to use OpenMP, I believe, currently.

Reply:
I believe most of the time one of `nx` / `nk` will be equal to 1. So to calculate multiple windows for same column, or same rolling window for multiple columns. In such cases we might easily lose threaded processing at all.
Merging those two loops won't result into "big fat" loop, as `nx` are `nk` are unlikely to be big numbers (of course there might some use cases to calculate 4000 columns on 4000 window sizes).
Before trying to merge I will look around, maybe there is some canonical way to handle that in openmp. If there isn't then best will be to see timings of two approaches.